### PR TITLE
Do not crash if the map is read only

### PR DESCRIPTION
**Pull request by HybridDog**
This add more robustness for read-only files. For example, when Minetest cannot save to a read-only map.sqlite, it shows an error message instead of crashing.
Features:
* Show errors instead of crashing for various read-only files
* If the map is read-only: nodes do not appear, or to be more precise, they appear more or less just temporarily due to node placement prediction. With this players who ignore the error message notice that something is wrong and they do not continue building for hours and suddenly notice that everything was in vain.

This PR addresses #7081

#### Remaining problems

* The auth.sqlite must be writeable
* The folder must me writeable
* Find out what https://github.com/minetest/minetest/pull/7544 has changed

### CODE REVIEW

File: src/map.cpp [hunk 1]
```diff
@@ -1931,6 +1931,13 @@ bool ServerMap::saveBlock(MapBlock *block)
 
 bool ServerMap::saveBlock(MapBlock *block, MapDatabase *db)
 {
+	// If the database is read-only, clear the modified flag without saving
+	static bool could_not_save = false;
```
Comment (nerzhul) [inline src/map.cpp]: it should not be only static but static thread_local as if you open a new world which is writeable you will block it

File: src/map.h [hunk 2]
```diff
@@ -291,6 +291,8 @@ class Map /*: public NodeContainer*/
 	void transforming_liquid_add(v3s16 p);
 
 	bool isBlockOccluded(MapBlock *block, v3s16 cam_pos_nodes);
+
+	bool is_readonly = false;
```
Comment (nerzhul) [inline src/map.h]: don't make it public, add setters and getters instead

File: src/serverenvironment.cpp [hunk 3]
```diff
@@ -526,26 +526,24 @@ void ServerEnvironment::saveLoadedPlayers()
...
-		}
+				player->getPlayerSAO()->extendedAttributesModified()))
+			savePlayer(player);
 	}
 }
 
 void ServerEnvironment::savePlayer(RemotePlayer *player)
 {
+	// don't try to save again if the database is read only
+	static bool could_not_save = false;
```
Comment (nerzhul) [inline src/serverenvironment.cpp]: same here

File: src/map.cpp (hunk 1 above)
Comment (HybridDog) [reply src/map.cpp]: fixed

**Comment by Lejo1** [comment]
A Setting to allow that blocks get setted, if the map is readonly would also be useful.

**Comment by HybridDog** [comment]
> A Setting to allow that blocks get setted, if the map is readonly would also be useful.

I assume after the map is unloaded, the nodes would disappear.

**Comment by Lejo1** [comment]
Ah Ok
Thanks

**Comment by twoelk** [comment]
would this allow to play a "demo-version" with a demo world from a cd without installing?
or would minetest still need to write to some file like updating player position, privs or such

**Comment by SmallJoker** [comment]
@twoelk Depending on what is set to read-only, the map can indeed be a "demo" map - until the write protection is removed (and the server restarted). But also the player database can be changed to read-only, which has the side-effect players can still be created when the "file" player backend is used and the "players" directory is writeable.

**Comment by HybridDog** [comment]
@SmallJoker rebased

**Comment by paramat** [comment]
@lhofhansl

**Comment by HybridDog** [comment]
Rebased.

**Comment by SmallJoker** [comment]
Needs rebase again. Please also fix the code style.

**Comment by HybridDog** [comment]
> Please also fix the code style.

Which code do you mean?

**Comment by SmallJoker** [comment]
@HybridDog I referred to the bracket positions that you fixed in 388c142. Thanks :)

File: builtin/game/forceloading.lua [hunk 4]
```diff
@@ -106,12 +100,19 @@ end)
...
 -- persists the currently forceloaded blocks to disk
 local function persist_forceloaded_blocks()
-	write_file(wpath.."/force_loaded.txt", blocks_forceloaded)
+	local f = io.open(wpath .. "/force_loaded.txt", "w")
+	if not f then
+		minetest.log("error",
+			"Could not open <worldpath>/force_loaded.txt for writing")
+		return
+	end
+	f:write(core.serialize(blocks_forceloaded))
```
Comment (SmallJoker) [inline builtin/game/forceloading.lua]: Safe write file could be used here.

**Comment by paramat** [comment]
The feature request issue was closed havong no core dev support after a year.
Does any core dev actually support the feature? No one has stated that (review does not necessarily imply it either) If not, perhaps this should be closed?

**Comment by pyrollo** [comment]
Globally this PR work, server does not crash anymore if map.sqlite is read only. I have two remarks:
- Before first message, dug/placed nodes are kept and not reverted. May be not important as first message comes quickly:
- It does not seem sufficient for starting a demo from a CD as server crashes if auth.sqlite or directory is read only. Another PR allowing run on RO auth base would be necessary.

**Comment by pyrollo** [comment]
I saw something has been done in order to be able to start with RO player DB.

Under Linux, merged master (5c4b560b6812bbdf7a9d3202c95bad5c9df97d66) into your branch.

Different cases:
- world directory RO, all files RO, auth.sqlite RO or RW;
- wolrd directory RW, all files RO;

Error:
```
2020-09-01 10:40:00: ACTION[Server]: singleplayer [127.0.0.1] joins game. List of players: singleplayer
terminate called after throwing an instance of 'DatabaseException'
  what():  : attempt to write a readonly database

Thread 13 "Server" received signal SIGABRT, Aborted.
[Switching to Thread 0x7fffcffff700 (LWP 1159211)]
__GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
50	../sysdeps/unix/sysv/linux/raise.c: Aucun fichier ou dossier de ce type.
```

Backtrace:
```
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007ffff6e7b859 in __GI_abort () at abort.c:79
#2  0x00007ffff7250951 in ?? () from /lib/x86_64-linux-gnu/libstdc++.so.6
#3  0x00007ffff725c47c in ?? () from /lib/x86_64-linux-gnu/libstdc++.so.6
#4  0x00007ffff725c4e7 in std::terminate() () from /lib/x86_64-linux-gnu/libstdc++.so.6
#5  0x00007ffff725c799 in __cxa_throw () from /lib/x86_64-linux-gnu/libstdc++.so.6
#6  0x0000555555da549b in Database_SQLite3::sqlite3_vrfy (this=0x555557389dd0, s=8, m="", r=101)
    at /home/pyrollo/dev/10-Minetest/06-MinetestMerge/minetest/src/database/database-sqlite3.h:113
#7  0x0000555555da5518 in Database_SQLite3::sqlite3_vrfy (this=0x555557389dd0, s=8, r=101, m="")
    at /home/pyrollo/dev/10-Minetest/06-MinetestMerge/minetest/src/database/database-sqlite3.h:118
#8  0x0000555555da3ff0 in AuthDatabaseSQLite3::saveAuth (this=0x555557389dd0, authEntry=...)
    at /home/pyrollo/dev/10-Minetest/06-MinetestMerge/minetest/src/database/database-sqlite3.cpp:705
#9  0x0000555555f3555e in ModApiAuth::l_auth_save (L=0x555556bc41d0) at /home/pyrollo/dev/10-Minetest/06-MinetestMerge/minetest/src/script/lua_api/l_auth.cpp:118
#10 0x00005555561e1ede in luaD_precall (L=0x555556bc41d0, func=0x555557069200, nresults=0)
    at /home/pyrollo/dev/10-Minetest/06-MinetestMerge/minetest/lib/lua/src/ldo.c:320
#11 0x00005555561f8577 in luaV_execute (L=0x555556bc41d0, nexeccalls=3) at /home/pyrollo/dev/10-Minetest/06-MinetestMerge/minetest/lib/lua/src/lvm.c:591
#12 0x00005555561e2191 in luaD_call (L=0x555556bc41d0, func=0x5555570690c0, nResults=1)
    at /home/pyrollo/dev/10-Minetest/06-MinetestMerge/minetest/lib/lua/src/ldo.c:378
#13 0x00005555561d9a74 in f_call (L=0x555556bc41d0, ud=0x7fffcfffe430) at /home/pyrollo/dev/10-Minetest/06-MinetestMerge/minetest/lib/lua/src/lapi.c:800
#14 0x00005555561e119b in luaD_rawrunprotected (L=0x555556bc41d0, f=0x5555561d9a3b <f_call>, ud=0x7fffcfffe430)
    at /home/pyrollo/dev/10-Minetest/06-MinetestMerge/minetest/lib/lua/src/ldo.c:116
#15 0x00005555561e259b in luaD_pcall (L=0x555556bc41d0, func=0x5555561d9a3b <f_call>, u=0x7fffcfffe430, old_top=48, ef=32)
    at /home/pyrollo/dev/10-Minetest/06-MinetestMerge/minetest/lib/lua/src/ldo.c:464
#16 0x00005555561d9b36 in lua_pcall (L=0x555556bc41d0, nargs=4, nresults=1, errfunc=2)
    at /home/pyrollo/dev/10-Minetest/06-MinetestMerge/minetest/lib/lua/src/lapi.c:821
#17 0x0000555555f1631b in ScriptApiBase::runCallbacksRaw (this=0x7fffd4014500, nargs=2, mode=RUN_CALLBACKS_MODE_FIRST, fxn=0x5555562a65f5 "on_joinplayer")
    at /home/pyrollo/dev/10-Minetest/06-MinetestMerge/minetest/src/script/cpp_api/s_base.cpp:277
#18 0x0000555555f2bce4 in ScriptApiPlayer::on_joinplayer (this=0x7fffd40144e8, player=0x7fffc0019e00, last_login=1598947259)
    at /home/pyrollo/dev/10-Minetest/06-MinetestMerge/minetest/src/script/cpp_api/s_player.cpp:163
#19 0x0000555555ee6529 in Server::handleCommand_ClientReady (this=0x555556bbcdb0, pkt=0x7fffcfffe860)
    at /home/pyrollo/dev/10-Minetest/06-MinetestMerge/minetest/src/network/serverpackethandler.cpp:416
#20 0x0000555556179052 in Server::handleCommand (this=0x555556bbcdb0, pkt=0x7fffcfffe860)
    at /home/pyrollo/dev/10-Minetest/06-MinetestMerge/minetest/src/server.cpp:1099
#21 0x0000555556165c51 in Server::ProcessData (this=0x555556bbcdb0, pkt=0x7fffcfffe860) at /home/pyrollo/dev/10-Minetest/06-MinetestMerge/minetest/src/server.cpp:1163
#22 0x0000555556164ec4 in Server::Receive (this=0x555556bbcdb0) at /home/pyrollo/dev/10-Minetest/06-MinetestMerge/minetest/src/server.cpp:997
#23 0x000055555615e232 in ServerThread::run (this=0x555556bac240) at /home/pyrollo/dev/10-Minetest/06-MinetestMerge/minetest/src/server.cpp:112
#24 0x0000555555faca5f in Thread::threadProc (thr=0x555556bac240) at /home/pyrollo/dev/10-Minetest/06-MinetestMerge/minetest/src/threading/thread.cpp:196
#25 0x0000555555fad9a9 in std::__invoke_impl<void, void (*)(Thread*), Thread*> (__f=@0x555557c54c50: 0x555555fac9ec <Thread::threadProc(Thread*)>)
    at /usr/include/c++/9/bits/invoke.h:60
#26 0x0000555555fad8f5 in std::__invoke<void (*)(Thread*), Thread*> (__fn=@0x555557c54c50: 0x555555fac9ec <Thread::threadProc(Thread*)>)
    at /usr/include/c++/9/bits/invoke.h:95
#27 0x0000555555fad845 in std::thread::_Invoker<std::tuple<void (*)(Thread*), Thread*> >::_M_invoke<0ul, 1ul> (this=0x555557c54c48) at /usr/include/c++/9/thread:244
#28 0x0000555555fad7e7 in std::thread::_Invoker<std::tuple<void (*)(Thread*), Thread*> >::operator() (this=0x555557c54c48) at /usr/include/c++/9/thread:251
#29 0x0000555555fad7b8 in std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (*)(Thread*), Thread*> > >::_M_run (this=0x555557c54c40)
    at /usr/include/c++/9/thread:195
#30 0x00007ffff7288cb4 in ?? () from /lib/x86_64-linux-gnu/libstdc++.so.6
#31 0x00007ffff75e4609 in start_thread (arg=<optimized out>) at pthread_create.c:477
#32 0x00007ffff6f78103 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

File: src/map.cpp [hunk 5]
```diff
@@ -1944,7 +1951,16 @@ bool ServerMap::saveBlock(MapBlock *block, MapDatabase *db)
 	o.write((char*) &version, 1);
 	block->serialize(o, version, true);
 
-	bool ret = db->saveBlock(p3d, o.str());
+	bool ret;
+	try {
+		ret = db->saveBlock(p3d, o.str());
+	} catch (DatabaseException &e) {
+		could_not_save = true;
```
Comment (rubenwardy) [inline src/map.cpp:1958]: This isn't necessarily only when the DB is readonly, I'm not sure `could_not_save` makes sense. For example, is there a DB exception on timeouts? Could certain backends cause timeouts when backups are running?

File: src/map.cpp [hunk 6]
```diff
@@ -1924,6 +1924,13 @@ bool ServerMap::saveBlock(MapBlock *block)
 
 bool ServerMap::saveBlock(MapBlock *block, MapDatabase *db)
 {
+	// If the database is read-only, clear the modified flag without saving
+	static thread_local bool could_not_save = false;
+	if (could_not_save) {
```
Comment (rubenwardy) [inline src/map.cpp:1929]: Why not check isReadonly on the database?

**Pull request closed without merging**